_posts/2017-11-06-jira-progressbar.md

```diff
@@ -0,0 +1,166 @@
+
+Zunächst benötigt man ein Custom Field vom Typ *Scripted Field*:
+
+![Select-a-Field-Type-Modal](/assets/images/jira-progressbar/jira-addfield.png)
+
+Im Anschluss wechselt man zu *Add-ons→Script Fields* und bearbeitet das gerade erzeugte Feld. Wichtig ist, das die Ausgabe auf *HTML* geändert wird.
+
+![Konfigurationsdialog für Scripted Field](/assets/images/jira-progressbar/jira-configurefield.png)
+
+## Script
+
+Bleibt nur noch, das Groovy-Skript als eigene Datei über den Serverpfad, oder als Inline-Script zu definieren.
+
+Als erstes suchen wir alle Vorgänge, die mittels *derives* beim aktuellen Vorgang verlinkt sind. Handelt es sich um einem Epic, so wird der Issue Key zwischengespeichert. Alle anderen Vorgänge werden auf deren Statuskategorie geprüft und der jeweilige Counter hochgezählt.
+
+~~~groovy
+import com.atlassian.jira.component.ComponentAccessor
+import com.atlassian.jira.issue.search.SearchProvider
+import com.atlassian.jira.jql.parser.JqlQueryParser
+import com.atlassian.jira.web.bean.PagerFilter
+
+def jqlQueryParser                = ComponentAccessor.getComponent(JqlQueryParser)
+def searchProvider                = ComponentAccessor.getComponent(SearchProvider)
+def issueManager                  = ComponentAccessor.getIssueManager()
+def user                          = ComponentAccessor.getJiraAuthenticationContext().getLoggedInUser()
+def statusCounter                 = ['New':0, 'In Progress':0, 'Complete':0]
+def epicslist                     = []
+
+def query                         = jqlQueryParser.parseQuery("issueFunction in linkedIssuesOf('issue = " + issue.getKey() + "', 'derives')")
+def results                       = searchProvider.search(query, user, PagerFilter.getUnlimitedFilter())
+
+results.getIssues().each {documentIssue ->
+    if (documentIssue.getIssueType().getName() == "Epic") {
+        epicslist.add(documentIssue.getKey())
+    } else {
+        statusCounter[documentIssue.getStatus().getStatusCategory().getName()] += 1
+    } // end if
+} // end each
+~~~
+
+Epics verhalten sich etwas anders, deswegen fragen wir darin liegende Tickets noch einmal über eine eigene JQL-Query ab. Dazu nutzen wir die Issue Keys von der ersten Abfrage. Auch hier aktualisieren wir die Counter wieder.
+
+~~~groovy
+if (epicslist.size() > 0) {
+    query                        = jqlQueryParser.parseQuery("issueFunction in issuesInEpics ('issuekey IN (" + epicslist.join(', ') + ")')")
+    results                      = searchProvider.search(query, user, PagerFilter.getUnlimitedFilter())
+
+    results.getIssues().each {documentIssue ->
+        statusCounter[documentIssue.getStatus().getStatusCategory().getName()] += 1
+    } // end each
+} // end if
+~~~
+
+Nun, da wir die benötigten Werte ermittelt haben, wird das ganze noch ausgegeben. Der Fortschrittsbalken wird mittels CSS-Inline-Grid erzeugt, sodass er zum Look-and-Feel von JIRA passt.
+
+~~~groovy
+/*
+return nothing of there are no linked issue
```

Comment:
@flaimo ìch glaub das sollte `if` statt `of` heißen?

Reply:
genau

Reply:
danke, fürs korrekturlesen.